### PR TITLE
rgw/rgw_rados: fix server side-copy orphans tail-objects

### PR DESCRIPTION
**Pull request by benhanokh**
Server-side copy copies the manifest from SRC->DST after incrementing the ref_count on the SRC tail-objects. The problem was that it uses the task-id as ref_tag and it got lost after the ref increment.When deleting the tail-objects from the COPY it passed in the obj_tag as ref_tag and since it was different from the one used during ref_inc it was ignored and the tail objects were never removed. This PR changes the ref_tag used for ref increment passing an MD5 of the bucket-name and the oid which is recreated when the COPY object is deleted

Fixes: https://tracker.ceph.com/issues/66286

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by benhanokh** [comment]
jenkins test api

**Comment by benhanokh** [comment]
jenkins test api

**Comment by benhanokh** [comment]
jenkins test make check

File: src/rgw/driver/rados/rgw_rados.cc [hunk 1]
```diff
@@ -4832,15 +4832,7 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
   RGWRados::Object dest_op_target(this, dest_bucket_info, obj_ctx, dest_obj);
   RGWRados::Object::Write write_op(&dest_op_target);
 
-  string tag;
-
-  if (ptag) {
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_rados.cc:4837]: if you arew not using `ptag` anymore,  why not moving it from the function prototype?

File: src/rgw/driver/rados/rgw_rados.cc [hunk 2]
```diff
@@ -4832,15 +4832,7 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
...
-    tag = *ptag;
-  }
-
-  if (tag.empty()) {
-    append_rand_alpha(cct, tag, tag, 32);
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_rados.cc:4842]: i guess that these tags are orphans anyway?
so, we are not creating a bigger issue when we will migrate RGW and change the way we generat the tags?

Comment (benhanokh) [reply src/rgw/driver/rados/rgw_rados.cc:4842]: the tag is only used for the ref_count, it is not set as id-tag (there is another code setting it regardless of the ptag value)

Comment (benhanokh) [reply src/rgw/driver/rados/rgw_rados.cc:4842]: BTW, not all flows pass in a valid tag, some are passing in a NULL value and then a random-alpha is generated

File: src/rgw/driver/rados/rgw_rados.cc [hunk 3]
```diff
@@ -5318,7 +5305,7 @@ int RGWRados::Object::complete_atomic_modification(const DoutPrefixProvider *dpp
     return 0;
   }
 
-  string tag = (state->tail_tag.length() > 0 ? state->tail_tag.to_str() : state->obj_tag.to_str());
+  string tag = state->get_obj_refcount_tag();
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_rados.cc:5308]: this had been using `state->tail_tag` which comes from `RGW_ATTR_TAIL_TAG`. i see that `copy_obj()` calls `attrs.erase(RGW_ATTR_TAIL_TAG)` in the ref-counting path. can we set that to `tag + '\0'` instead?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5308]: wait, that should be happening already in copy_obj() -> write_meta() -> prepare_atomic_modification()

Comment (benhanokh) [reply src/rgw/driver/rados/rgw_rados.cc:5308]: The PR kept the state->tail_tag unmodified and it is still used as the ref_tag on the SRC object.
The copy_obj() erase the TAIL_TAG on the target destination as an indication that it is was produced by copy and in that case will use the ID_TAG as ref_tag (which is wrong, as we used something else )

File: src/rgw/driver/rados/rgw_rados.cc [hunk 4]
```diff
@@ -4909,7 +4896,7 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
 
   write_op.meta.data = &first_chunk;
   write_op.meta.manifest = pmanifest;
-  write_op.meta.ptag = &tag;
+  write_op.meta.ptag = 0;
```
Comment (ronen-fr) [inline src/rgw/driver/rados/rgw_rados.cc:4899]: nullptr?

File: src/rgw/rgw_obj_types.h [hunk 5]
```diff
@@ -27,7 +27,7 @@
 
 #include "common/dout.h"
 #include "common/Formatter.h"
-
+#include "common/ceph_crypto.h"
```
Comment (ronen-fr) [inline src/rgw/rgw_obj_types.h:30]: should probably keep the blank line

File: src/rgw/driver/rados/rgw_rados.cc (hunk 3 above)
Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5308]: > The copy_obj() erase the TAIL_TAG on the target destination as an indication that it is was produced by copy and in that case will use the ID_TAG as ref_tag (which is wrong, as we used something else )

copy_obj() erases RGW_ATTR_TAIL_TAG, but later adds it back in prepare_atomic_modification(). it looks like that should correctly store the new ref_tag in both RGW_ATTR_ID_TAG and RGW_ATTR_TAIL_TAG. are you sure that's not happening?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5308]: the target object _should_ be atomic here, and i'd expect `s->object->set_atomic()` in [RGWCopyObj::verify_permission()](https://github.com/ceph/ceph/blob/cac7dcd6/src/rgw/rgw_op.cc#L5397) to make it so

Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5308]: ok, the problem is that we're calling s->src_object->copy_object() which calls copy_obj() with src_object's RGWObjectCtx. s->object->set_atomic() is setting the atomic flag in the RGWObjectCtx of s->object

cc @dang, i think this was a regression from https://github.com/ceph/ceph/pull/44616. prior to that, we use the same `s->obj_ctx` for both objects

**Comment by dang** [comment]
@benhanokh I posted https://github.com/ceph/ceph/pull/58320 that should fix the obj_ctx issue for copy_object().  Can you review and test it please?

**Comment by cbodley** [comment]
thanks @dang, i tested https://github.com/ceph/ceph/pull/58320 in a vstart cluster with:
```
$ s3cmd mb s3://testbucket
$ s3cmd put 128m.iso s3://testbucket
$ s3cmd cp s3://testbucket/128m.iso s3://testbucket/128m.bak
$ bin/radosgw-admin object stat --bucket testbucket --object 128m.iso
$ bin/radosgw-admin object stat --bucket testbucket --object 128m.bak
```
before these changes, the second `object stat` command output did not include "user.rgw.tail_tag". it now does, and that tag is different from the attr returned by 128m.iso as expected
```
$ s3cmd rm s3://testbucket/128m.iso
$ s3cmd rm s3://testbucket/128m.bak
$ bin/radosgw-admin gc process --include-all
$ bin/rados -p default.rgw.buckets.data ls
```
after deleting both copies and running gc, 'rados ls' shows no orphaned objects 👍 

the test_rgw_orphan_list.sh changes in this pr look like a great way to validate the fix. @benhanokh if you're satisfied with Dan's approach, would you mind pulling that commit into this pr so we can get it through qa?

**Comment by benhanokh** [comment]
@cbodley, assuming you need the TAIL-TAG in the OBJ-COPY then Dan's approach will work better.
Do you think the ref-tag could be hashed to something smaller (like I do in my code) or there is a reason we need 64 bytes long string tags?

In any case Dan's PR needs to add test cases for deleting all copies and making sure no orphans are left behind (can take the test cases from my PR)

**Comment by cbodley** [comment]
> Do you think the ref-tag could be hashed to something smaller (like I do in my code) or there is a reason we need 64 bytes long string tags?

i think this would need further investigation. a hash based on bucket+object name might not do what we want in the case of overwrites:
> upload object bucket/A
copy bucket/A -> bucket/B
copy bucket/A -> bucket/B

in that case, the tail objects from bucket/A would get two refs with the same bucket/B tag. but i _think_ the intent is for each ref to have a different tag so that garbage collection is idempotent - if gc crashes in the middle of derefs, we'll restart and deref the same tail objects again. if we use a separate tag for each ref, then the second deref will be ignored

**Comment by cbodley** [comment]
> In any case Dan's PR needs to add test cases for deleting all copies and making sure no orphans are left behind (can take the test cases from my PR)

@benhanokh i was thinking that you could take Dan's commit into this pr since we want your test cases and `rados refcount` command. but the opposite would also work - please coordinate with @dang

**Comment by benhanokh** [comment]
> > In any case Dan's PR needs to add test cases for deleting all copies and making sure no orphans are left behind (can take the test cases from my PR)
> 
> @benhanokh i was thinking that you could take Dan's commit into this pr since we want your test cases and `rados refcount` command. but the opposite would also work - please coordinate with @dang

done

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check

**Comment by cbodley** [comment]
@benhanokh can you please remove your own changes from the branch (ie squash "rollback changes" into the ealier commits) and cherry-pick Dan's commit to preserve his commit title, authorship, and signoff?

**Comment by cbodley** [comment]
in the meantime, i ran this through the rgw suite: https://pulpito.ceph.com/cbodley-2024-07-10_21:34:12-rgw-wip-66286-distro-default-smithi/

in https://qa-proxy.ceph.com/teuthology/cbodley-2024-07-10_21:34:12-rgw-wip-66286-distro-default-smithi/7795944/teuthology.log, test_rgw_orphan_list.sh failed after an osd crashed with `No space left on device`
```
2024-07-11T00:43:54.835 INFO:tasks.ceph.osd.0.smithi117.stderr:2024-07-11T00:43:54.834+0000 7fcf95a40640 -1 bdev(0x559e0aec2e00 /var/lib/ceph/osd/ceph-0/block) _aio_thread got r=-28 ((28) No space left on device)
2024-07-11T00:43:54.836 INFO:tasks.ceph.osd.0.smithi117.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4914-ga38a2372/rpm/el9/BUILD/ceph-19.0.0-4914-ga38a2372/src/blk/kernel/KernelDevice.cc: In function 'void KernelDevice::_aio_thread()' thread 7fcf95a40640 time 2024-07-11T00:43:54.835438+0000
2024-07-11T00:43:54.836 INFO:tasks.ceph.osd.0.smithi117.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4914-ga38a2372/rpm/el9/BUILD/ceph-19.0.0-4914-ga38a2372/src/blk/kernel/KernelDevice.cc: 666: ceph_abort_msg("Unexpected IO error. This may suggest a hardware issue. Please check your kernel log!")
2024-07-11T00:43:54.836 INFO:tasks.ceph.osd.0.smithi117.stderr: ceph version 19.0.0-4914-ga38a2372 (a38a23721a6485d87d9a339a64f3539a1382b378) squid (dev)
2024-07-11T00:43:54.836 INFO:tasks.ceph.osd.0.smithi117.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xc6) [0x559e08ccef8b]
```
earlier in the log there were a bunch of s3cmd processes killed, i'm not sure whether that's expected or related:
```
2024-07-10T22:13:03.901 INFO:tasks.workunit.client.0.smithi117.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 39232 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-10T22:14:15.379 INFO:tasks.workunit.client.0.smithi117.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 39315 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-10T22:14:23.402 INFO:tasks.workunit.client.0.smithi117.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43339 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
```

**Comment by benhanokh** [comment]
@cbodley,
The tests are writing "huge objects" of 5GB each.
My tests added 3x3 s3cmd put for it with a total of 45GB (in addition to the 30GB written by the previous 2 tests)
We perform recursive delete after each test which should limit the used space, but the GC is only called after all the tests.

Maybe we should call GC after each test to limit the used space?

**Comment by cbodley** [comment]
> Maybe we should call GC after each test to limit the used space?

sounds good

**Comment by benhanokh** [comment]
jenkins test make check

**Comment by cbodley** [comment]
https://pulpito.ceph.com/cbodley-2024-07-16_17:46:19-rgw-wip-66286-distro-default-smithi/

shows lots of s3test failures because the branch is missing some fixes from main:
```
2024-07-16T18:32:47.041 INFO:teuthology.orchestra.run.smithi096.stdout:=========================== short test summary info ============================
2024-07-16T18:32:47.041 INFO:teuthology.orchestra.run.smithi096.stdout:FAILED s3tests_boto3/functional/test_s3.py::test_multipart_single_get_part - ...
2024-07-16T18:32:47.041 INFO:teuthology.orchestra.run.smithi096.stdout:FAILED s3tests_boto3/functional/test_s3.py::test_object_checksum_sha256 - Key...
2024-07-16T18:32:47.041 INFO:teuthology.orchestra.run.smithi096.stdout:FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_sha256 - ...
2024-07-16T18:32:47.042 INFO:teuthology.orchestra.run.smithi096.stdout:FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_3parts - ...
2024-07-16T18:32:47.042 INFO:teuthology.orchestra.run.smithi096.stdout:FAILED s3tests_boto3/functional/test_s3.py::test_post_object_upload_checksum
```
those should go away with a rebase

but test_rgw_orphan_list.sh failed with what looks like ENOSPC again: http://qa-proxy.ceph.com/teuthology/cbodley-2024-07-16_17:46:19-rgw-wip-66286-distro-default-smithi/7804039/teuthology.log

**Comment by cbodley** [comment]
@ivancich regarding your question in https://tracker.ceph.com/issues/66286#note-6, we're trying to get these orphan test cases passing. they keep causing osds to fill up and crash with ENOSPC, despite the addition of `gc process` commands. i also noticed that the latest run of test_rgw_orphan_list.sh took over 3 hours before failing this way

can you think of other ways to reduce the storage requirements of this testing? does it need to operate at this scale to provide sufficient regression test coverage?

separately, i've noticed a lot of odd errors from the [teuthology logs](https://qa-proxy.ceph.com/teuthology/cbodley-2024-07-16_17:46:19-rgw-wip-66286-distro-default-smithi/7804039/teuthology.log) - any idea what's happening here?
```
2024-07-16T18:24:56.783 INFO:tasks.workunit.client.0.smithi122.stdout:Bucket 's3://incomplete-mp-bkt-1/' created
2024-07-16T18:25:05.867 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 38928 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:16.144 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 39011 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:24.154 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43035 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:32.169 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43050 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:40.180 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43065 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:48.192 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43080 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:26:56.210 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43095 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:27:04.211 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43110 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:27:12.231 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43125 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:27:20.262 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43140 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:27:28.264 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43155 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
2024-07-16T18:27:36.255 INFO:tasks.workunit.client.0.smithi122.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_orphan_list.sh: line 158: 43170 Killed                  s3cmd --config=${s3config} put $local_file s3://${remote_bkt}/${remote_obj} --progress --multipart-chunk-size-mb=5 > $fifo
```

**Comment by benhanokh** [comment]
My machine can't test latest code because it is centos 8.
I'm trying to get it moved to centos 9.
In the meanwhile I can add more calls to GC between tests 4 and 5, and between tests 6 and 7

**Comment by cbodley** [comment]
> Maybe I can also nreduce the huge-obj size from 5GB down to 1GB?

is there a reason to use 'huge' objects, other than the fact that they contain at least one tail object? maybe 5MB would suffice there

**Comment by benhanokh** [comment]
This is how the original tests were used
We can probably scale them down to 50MB giving us a dozen tail objects without overloading the system

**Comment by benhanokh** [comment]
My machine is still waiting for an upgrade from centos 8 so I was unable to test anything.
This is a single line change shrinking the object-size used for orphan-list test by a factor of 100 from 5100MB -> 51MB.

I hope it will solve the out-of-space problem.
If not we will need to wait until my machine is upgraded and I acn run the tests locally

**Comment by cbodley** [comment]
yay, [latest job](https://qa-proxy.ceph.com/teuthology/cbodley-2024-07-23_16:02:01-rgw:tools-wip-66286-distro-default-smithi/7814091/teuthology.log) succeeded with `0 potential orphans found out of a possible 4319 (0%).`

> i also noticed that the latest run of test_rgw_orphan_list.sh took over 3 hours before failing this way

better still, the total runtime was reduced to 0:26:55

**Comment by cbodley** [comment]
also scheduled the updated test case against packages from main to verify that it catches the refcount issue:
> $ teuthology-suite -m smithi -s rgw:tools --ceph-repo https://github.com/ceph/ceph.git -c main -S 86385b1daf5235400bbcd34a02515d168bde88e2 --suite-repo https://github.com/ceph/ceph-ci.git --suite-branch wip-66286 -p 75

pending in https://pulpito.ceph.com/cbodley-2024-07-23_16:43:49-rgw:tools-main-distro-default-smithi/

**Comment by cbodley** [comment]
> also scheduled the updated test case against packages from main to verify that it catches the refcount issue:

that did indeed detect the refcount orphans: `28 potential orphans found out of a possible 4347 (0%).` 👍

**Comment by cbodley** [comment]
jenkins test api

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed the rgw suite in https://pulpito.ceph.com/cbodley-2024-07-30_11:48:11-rgw-wip-cbodley-testing-distro-default-smithi/

we want this on squid by the next rc (which branches tomorrow) so i'm not blocking this on @ronen-fr's review

File: src/tools/rados/rados.cc [hunk 6]
```diff
@@ -2750,8 +2752,30 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+      if (!oref.retired_refs.empty()) {
+	cout << "--------------------------------------" << std::endl;
+	for (const auto & ref : oref.retired_refs) {
+	  cout << "retired_refs::" << ref << std::endl;
+	}
+      }
+    }
+    else {
+      string s(bl.c_str(), bl.length());
+      cout << s << std::endl;
```
Comment (mchangir) [inline src/tools/rados/rados.cc:2777]: The last `std::endl` is superfluous and is breaking cephfs tests.
The `std::endl` did not exist earlier.
Is there any specific requirement to have an EOL here or could we change it to `std::flush` ?

Comment (cbodley) [reply src/tools/rados/rados.cc:2777]: sorry! whatever it takes to fix the cephfs test. reverting to `cout << s;` should be fine

Comment (cbodley) [reply src/tools/rados/rados.cc:2777]: looks like the rados suite caught this too, see https://github.com/ceph/ceph/pull/59025

**Pull request merged**